plugins/ecs-bridge/engine/engine.go

```diff
@@ -0,0 +1,234 @@
+// It defines all the operations performed during the execution of the
+// plugin
+type Engine interface {
+	CreateBridge(bridgeName string, mtu int) (*netlink.Bridge, error)
+	CreateVethPair(netnsName string, bridge *netlink.Bridge, mtu int, interfaceName string) (*current.Interface, *current.Interface, error)
+	RunIPAMPluginAdd(plugin string, netConf []byte) (*current.Result, error)
+	ConfigureContainerVethInterface(netnsName string, result *current.Result, interfaceName string) error
+	ConfigureBridge(result *current.Result, bridge *netlink.Bridge) error
+	GetInterfaceIPV4Address(netnsName string, interfaceName string) (string, error)
+	RunIPAMPluginDel(plugin string, netconf []byte) error
+	DeleteVeth(netnsName string, interfaceName string) error
+}
+
+type engine struct {
+	netLink netlinkwrapper.NetLink
+	ns      cninswrapper.NS
+	ip      cniipwrapper.IP
+	ipam    cniipamwrapper.IPAM
+}
+
+// New creates a new Engine object
+func New() Engine {
+	return &engine{
+		netLink: netlinkwrapper.NewNetLink(),
+		ns:      cninswrapper.NewNS(),
+		ip:      cniipwrapper.New(),
+		ipam:    cniipamwrapper.New(),
+	}
+}
+
+// CreateBridge creates the bridge if needed
+func (engine *engine) CreateBridge(bridgeName string, mtu int) (*netlink.Bridge, error) {
+	bridge := &netlink.Bridge{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:   bridgeName,
+			MTU:    mtu,
+			TxQLen: -1,
+		},
+	}
+
+	err := engine.netLink.LinkAdd(bridge)
+	if err != nil && err != syscall.EEXIST {
+		return nil, errors.Wrapf(err,
+			"bridge create: unable to add bridge interface %s", bridgeName)
+	}
+
+	bridgeLink, err := engine.netLink.LinkByName(bridgeName)
+	if err != nil {
+		return nil, errors.Wrapf(err,
+			"bridge create: unable to lookup the bridge interface %s", bridgeName)
+	}
+	bridge, ok := bridgeLink.(*netlink.Bridge)
+	if !ok {
+		return nil, errors.Wrapf(err,
+			"bridge create: interface named %s already exists, but is not a bridge", bridgeName)
+	}
+
+	if err := engine.netLink.LinkSetUp(bridge); err != nil {
+		return nil, errors.Wrapf(err,
+			"bridge create: unable to bring up the bridge interface %s", bridgeName)
+	}
+
+	return bridge, nil
+}
+
+// CreateVethPair creates the veth pair to attach the container to
+// the bridge
+func (engine *engine) CreateVethPair(netnsName string, bridge *netlink.Bridge, mtu int, interfaceName string) (*current.Interface, *current.Interface, error) {
+	setupContext := newCreateVethPairContext(
+		interfaceName, mtu, engine.ip)
+
+	err := engine.ns.WithNetNSPath(netnsName, setupContext.run)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	hostVethName := setupContext.hostVethName
+	containerInterfaceResult := setupContext.containerInterfaceResult
+
+	hostVeth, err := engine.netLink.LinkByName(hostVethName)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err,
+			"bridge create veth pair: unable to look up host veth interface %s", hostVethName)
+	}
+
+	err = engine.netLink.LinkSetMaster(hostVeth, bridge)
```

Comment:
I think `LinkSetMaster` is fairly obvious without needing our own abstractions. Can you please elaborate what's poor about this implementation?